### ISSUE TITLE
Parameter index dependency

### ISSUE THREAD

**Issue by sverhoeven** [i-VRESSE/workflow-builder#59]
The index of some array parameters like `mol_shape` in rigidbody node tell to which molecule in global molecules parameter they belong.

For example `mol_shape=[false, false, true]` means that the third molecule should be treated as an shape.

It would be nice if 
1. the `maxitems` of mol_shape was somehow equal to the number of molecules the user uploaded.

For bonus points it would also be nice if
1. the parameters which needs a chain be picked for a certain molecule would be limited to the choices of the chains available in the molecule.
2. the parameters which need residue numbers would be limited to the ones available in the molecule.
3.  the parameters which needs a segment ID (last column in atom lines of PDB) be picked for a certain molecule would be limited to the choices of the segment IDs available in the molecule.
To find chains, residues, segment ids could use https://www.npmjs.com/package/parse-pdb or https://github.com/molstar/molstar

**Comment by amjjbonvin** [comment]
Or our own pdb-tools!
> To find chains and residues could use https://www.npmjs.com/package/parse-pdb <https://www.npmjs.com/package/parse-pdb> or https://github.com/molstar/molstar <https://github.com/molstar/molstar>

**Comment by sverhoeven** [comment]
A generic way could be for in the nodes JSON schema add a custom key `maxitemsFrom` with a [pointer](https://tools.ietf.org/html/rfc6901) to another node or global. The format could be `< node name | global>/<JSON pointer>`. For example `global/molecules`. 

1. Catalog generator inserts `maxitemsFrom` in schema when array is tied to a molecule
2. During app catalog loading the schemas are grepped for `maxitemsFrom` occurrences
3. Each time the length of value of the parameter of the pointer changes then the JSON schema of the node with the `maxitemsFrom` key should be regenerated with an updated `maxItems`.

**Comment by sverhoeven** [comment]
To do the residues/chain/segment id filtering we need to perform molecule parsing and inject ranges and enums into the schema of parameters that need filtering. 

We could define something like
```
c2sym:
  type: array
  maxitemsFrom: global/molecules
  items:
    type: array
    title: Segments
    items:
      type: object
      properties:
        sta1:
          type: number
          title: Residue number of first residue in the first C2 symmetry segment
          format: residue
        seg1:
          type: string
          title: Segment ID of first C2 segment
          format: segmentid
```
Then the form widgets should somehow render a select box with all choices for that molecule and the validator should check the selected value is valid.

**Comment by rvhonorato** [comment]
> Or our own pdb-tools!

That would be great, but the builder runs on typescript, not trivial to integrate it with python (there's no backend executing code in a server somewhere, everything runs in the browser).

**Comment by sverhoeven** [comment]
Shapes that can have chain or residue formats in them

### 1. array>object>scalar

```yaml
      c2sym:
        type: array
        items:
          type: object
          properties:
            sta1:  # c2sym_sta1_1
              type: number
              format: residue
```
(c?sym, s3sym, ncs have similar shapes)

### 2. array>object>array>scalar
```yaml
      mol:
        type: array
        items:
          type: object
          properties:
            hisd:  # mol.hisd_1
              type: array
              items:
                title: HISD residue number
                type: number
                format: residue
```

### 3. array>array>object>scalar

```yaml
      seg:
        type: array
        items:
          type: array
          items:
            type: object
            properties:
              sta: # seg_sta_1_1
                type: number
                format: residue
```
(fle and rair have similar shape)

In shape 2 and 3 the first array is indexed by the uploaded molecules.

Shape 1 (c2sym) looks like a bag of constraints where the array index is not tied to a certain molecule.
What should the residues and segment ids be restricted to? 
1. A union of all residues in all molecules
2. Residues from the smallest molecule
3. Residues from the biggest molecule
4. Not restricted at all
5. Something else

**Comment by sverhoeven** [comment]
After discussion with @rvhonorato , it became clear that haddock expects each molecule file to have a single chain and that chain names are not duplicated. 

So shape 1 has an additional property to select chain. Where as shape 2 and 3 the array index determines the molecule. This chain property gives the same protein as the array index method.

**Comment by sverhoeven** [comment]
Fixed in #62 and some unresolved corner case has been separated into #78
